### ISSUE TITLE
AsyncMysqlClient::adoptConnection & extended PDO class

### ISSUE THREAD

**Issue by aewing** [facebook/hhvm#6198]
What are the intended argument options for AsyncMysqlClient::adoptConnection?

I am trying to instantiate a new AsyncMysqlClient based off of a class in hack that extends 
PDO, but hhvm throws a fatal error: "Unexpected object type [my\classname]"

**Comment by JoelMarcey** [comment]
@aewing Funny you should ask that. I am documenting this currently as part of our documentation revamp, and I am running into the same problem. I think it doesn't support PDO or `mysqli` connections; just `mysql` maybe? And even that, I am not sure it is working correctly yet.

That said, I think the recommended way to instead use an `AsyncMysqlConnectionPool`. I believe we are trying to steer everyone in that direction.

cc @fredemmott

**Comment by aewing** [comment]
If it could support existing PDO objects that would be incredibly beneficial to my use case, I'm hoping to allow the end user to to "swap" into an awaitable interface at will. If not I suppose I'll go back to the drawing board.

**Comment by aewing** [comment]
I suppose it would make more sense in my situation for the AsyncMysqlConnectionPool to have some way to extract the connection parameters or duplicate my PDO instance. Is this something I shouldn't hold my breath for?

**Comment by fredemmott** [comment]
- this function requires a (deprecated) mysql resource

TL;DR for the rest: the best solution is to modify whatever gives you PDO objects to also have a pool, and to be able to give you async mysql clients

- after adopting the resource, it is no longer valid for use in any other context; so, really whatever kind of connection you already have, you want to get a new instance to pass to this function anyway

In general, uses of this function are a smell, and tend to create failures at a distance. I think we should deprecate it.

To use this, you would need to have a factory for new mysql connection objects; that factory presumably has the connection details, and could host a pool and also provide a way to get new async connections from that pool?

I also don't think the pool should be able to copy the connection details.
 - if you can make new connections at will, you presumably have the parameters handy
 - if you can't, spawning new connections from a single one is probably breaking another components' encapsulation

**Comment by fredemmott** [comment]
I'm fairly likely to wontfix this and mark adoptConnection as deprecated. Going to think about it for a while more.

**Comment by lexidor** [comment]
I am going over old issues on this repository, to see which ones apply to the current versions of hhvm.

https://hhvm.com/blog/2022/02/24/hhvm-4.150.html

This method was removed in 4.150.
